include/aspect/particle/particle.h

```diff
@@ -21,119 +21,141 @@
-                      const double &new_id);
+        Particle (const Point<dim> &new_loc,
+                  const particle_index &new_id);
+
+        /**
+         * Constructor for Particle, creates a particle from a data vector.
+         * This constructor is usually called after sending a particle to a
+         * different process.
+         *
+         * @param[in,out] begin_data First data component.
+         * @param[in] data_len Number of components of the begin_data vector
+         * that will be read in by this particle.
+         */
+        Particle (const void *&begin_data,
+                  const unsigned int data_len);
 
         /**
-         * Destructor for BaseParticle
+         * Destructor for Particle
          */
         virtual
-        ~BaseParticle ();
+        ~Particle ();
 
         /**
          * Get the number of doubles required to represent this particle for
-         * communication.
+         * communication. This includes the base properties like position
+         * and id as well as the additional user requested properties.
          *
          * @return Number of doubles required to represent this particle
          */
-        static unsigned int
-        data_len ();
+        unsigned int
+        data_len () const;
 
         /**
-         * Read the particle data from the specified vector of doubles.
-         *
-         * @param [in] data The vector of double data to read from.
-         * @param [in] pos The position in the data vector to start reading
-         * from.
-         * @return The position in the vector of the next unread double.
-         */
-        virtual unsigned int read_data(const std::vector<double> &data, const unsigned int &pos);
+          * Set the number of doubles required to represent this particle for
+          * communication. This includes the base properties position
+          * and id as well as the additional user requested properties.
+          *
+          * @param [in] data_len Number of doubles to represent this particle
+          */
+        void
+        set_data_len (const unsigned int data_len);
 
         /**
-         * Write particle data to a vector of doubles.
+         * Write particle data into a data array. The array is expected
+         * to be large enough to take the data, and the void pointer should
+         * point to the first element in which the data should be written. This
+         * function is meant for serializing all particle properties and
+         * afterwards de-serializing the properties by calling the appropriate
+         * constructor Particle(void *&data, const unsigned int data_len);
          *
-         * @param [in,out] data The vector of doubles to write integrator data
-         * into.
+         * @param [in,out] data The memory location to write particle data
+         * into. This pointer points to the begin of the memory, in which the
+         * data will be written.
```

Comment:
I think you want to say that it points to this memory location *before the call* and that after the call the function will have advanced it somehow. By how much will it advance? By `data_len()*8`? If that's the rule, then my preference would actually be if the function only took the pointer by value, and that the increment happens at the calling side. Functions that modify scalar arguments passed by reference are often hard to understand.

If the rule above does not hold, how about *returning* the incremented variable instead?